pyadjoint/overloaded_type.py

```diff
@@ -239,6 +239,17 @@ def _ad_will_add_as_output(self):
 
         """
         return True
+    
+    def _ad_petsc_vec(self):
```

Comment:
I'd prefer `_ad_to_petsc` and `_ad_from_petsc`, as I think that is all that is needed for the TAO interface, and that lets Firedrake use context managers in the implementation. That does mean there would be no copy-free access of the `PETScVec`.

Reply:
Actually this might not work, as the `IS` objects are still needed